src/unix/fs.c

```diff
@@ -766,6 +770,130 @@ static ssize_t uv__fs_write(uv_fs_t* req) {
   return r;
 }
 
+static ssize_t uv__fs_copyfile(uv_fs_t* req) {
+#if defined(__APPLE__)
+  /* On macOS, use the native copyfile(3). */
+  copyfile_flags_t flags;
+
+  flags = COPYFILE_ALL;
+
+  if (req->flags & UV_FS_COPYFILE_EXCL)
+    flags = flags | COPYFILE_EXCL;
+
+  return copyfile(req->path, req->new_path, NULL, flags);
+#else
+  uv_fs_t fs_req;
+  uv_file src;
+  uv_file dst;
+  uv_buf_t buf;
+  char data[4096];
+  struct stat statsbuf;
+  int offset;
+  int num_read;
+  int dst_flags;
+  int result;
+  int err;
+
+  dst = -1;
+  err = 0;
+
+  /* Open the source file. */
+  src = uv_fs_open(req->loop, &fs_req, req->path, O_RDONLY, 0, NULL);
+  uv_fs_req_cleanup(&fs_req);
+
+  if (src < 0) {
+    err = src;
+    goto out;
+  }
+
+  /* Get the source file's mode. */
+  err = fstat(src, &statsbuf);
+
+  if (err != 0) {
+    err = -errno;
+    goto out;
+  }
+
+  dst_flags = O_WRONLY | O_CREAT;
+
+  if (req->flags & UV_FS_COPYFILE_EXCL)
+    dst_flags = dst_flags | O_EXCL;
+
+  /* Open the destination file. */
+  dst = uv_fs_open(req->loop,
+                   &fs_req,
+                   req->new_path,
+                   dst_flags,
+                   statsbuf.st_mode,
+                   NULL);
+  uv_fs_req_cleanup(&fs_req);
+
+  if (dst < 0) {
+    err = dst;
+    goto out;
+  }
+
+  /* Setup the buffer and read from the source file in a loop. */
+  buf.base = data;
+  offset = 0;
+
+  do {
+    buf.len = sizeof(data);
+    num_read = uv_fs_read(req->loop, &fs_req, src, &buf, 1, offset, NULL);
+    uv_fs_req_cleanup(&fs_req);
+
+    if (num_read > 0) {
+      buf.len = num_read;
+      err = uv_fs_write(req->loop, &fs_req, dst, &buf, 1, offset, NULL);
+      uv_fs_req_cleanup(&fs_req);
+
+      if (err < 0)
+        goto out;
+
+      assert(err == num_read);
+      offset += num_read;
+    } else if (num_read < 0) {
+      err = num_read;
+      goto out;
+    }
+  } while (num_read > 0);
+
+out:
+  if (err < 0)
+    result = err;
+  else
+    result = 0;
+
+  /* Close the source and destination files. */
+  if (src >= 0) {
+    err = uv_fs_close(NULL, &fs_req, src, NULL);
+    uv_fs_req_cleanup(&fs_req);
+
+    /* Don't overwrite any existing errors. */
+    if (err != 0 && result == 0)
+      result = err;
+  }
+
+  if (dst >= 0) {
+    err = uv_fs_close(NULL, &fs_req, dst, NULL);
+    uv_fs_req_cleanup(&fs_req);
+
+    /* Don't overwrite any existing errors. */
+    if (err != 0 && result == 0)
+      result = err;
+
+    /* Remove the destination file if something went wrong. */
+    if (result != 0) {
+      err = uv_fs_unlink(NULL, &fs_req, req->new_path, NULL);
```

Comment:
It's probably academic but there is a race window between closing and unlinking; the file might have been changed or replaced by another process.

I don't know of a cross-platform way of doing it race-free but an approach I've used in the past is to create a file with a random name in the target directory, write out the data, then `rename(2)` it.

(On Linux, you can open the file with `O_TMPFILE` and then link `/proc/self/fd/<fd>` into the file system afterwards but that isn't portable and requires a relatively recent kernel.)

Perhaps we could start by simply adding a note to the documentation. :-)